AIPscan/Data/data.py

```diff
@@ -224,6 +254,84 @@ def derivative_overview(storage_service_id):
     return report
 
 
+def _format_versions_count_query(storage_service_id, start_date=None, end_date=None):
+    """Fetch information from database on format versions.
+
+    :param storage_service_id: Storage Service ID.
+    :start_date: Optional inclusive AIP creation start date
+    (datetime.datetime object).
+    :end_date: Optional inclusive AIP creation end date
+    (datetime.datetime object).
+
+    :returns: SQLAlchemy query results
+    """
+    PUID = "puid"
+    FILE_FORMAT = "file_format"
+    FORMAT_VERSION = "format_version"
+    FILE_COUNT = "file_count"
+    FILE_SIZE = "total_size"
+
+    start_date, end_date = _prepare_datetimes_for_query(start_date, end_date)
+
+    return (
+        db.session.query(
+            File.puid.label(PUID),
+            File.file_format.label(FILE_FORMAT),
+            File.format_version.label(FORMAT_VERSION),
+            db.func.count(File.id).label(FILE_COUNT),
+            db.func.sum(File.size).label(FILE_SIZE),
+        )
+        .join(AIP)
+        .join(StorageService)
+        .filter(StorageService.id == storage_service_id)
+        .filter(File.file_type == FileType.original.value)
+        .filter(AIP.create_date >= start_date)
+        .filter(AIP.create_date < end_date)
+        .group_by(File.puid)
+        .order_by(db.func.count(File.id).desc(), db.func.sum(File.size).desc())
+    )
+
+
+def format_versions_count(storage_service_id, start_date=None, end_date=None):
+    """Return a summary of format versions in Storage Service.
+
+    :param storage_service_id: Storage Service ID.
+    :start_date: Optional inclusive AIP creation start date
+    (datetime.datetime object).
+    :end_date: Optional inclusive AIP creation end date
+    (datetime.datetime object).
+
+    :returns: "report" dict containing following fields:
+        report["StorageName"]: Name of Storage Service queried
+        report["FormatVersions"]: List of result files ordered desc by size
+    """
+    report = {}
+    report[FIELD_FORMAT_VERSIONS] = []
+    report[FIELD_STORAGE_NAME] = None
+
+    storage_service = _get_storage_service(storage_service_id)
+    if storage_service is not None:
+        report[FIELD_STORAGE_NAME] = storage_service.name
+
+    versions = _format_versions_count_query(storage_service_id, start_date, end_date)
+
+    for version in versions:
+        version_info = {}
+
+        version_info[FIELD_PUID] = version.puid
+        version_info[FIELD_FORMAT] = version.file_format
+        try:
+            version_info[FIELD_VERSION] = version.format_version
+        except AttributeError:
+            pass
+        version_info[FIELD_COUNT] = version.file_count
+        version_info[FIELD_SIZE] = version.total_size
```

Comment:
Ah, it looks like maybe we can test for none here? (see template comment)

Reply:
Yes I think you're right!